leakcanary-deobfuscation-gradle-plugin/src/main/java/com/squareup/leakcanary/deobfuscation/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  private fun createLeakCanaryExtension(project: Project): LeakCanaryDeobfuscationExtension {
+    return project.extensions.create("leakCanary", LeakCanaryDeobfuscationExtension::class.java)
+  }
+
+  private fun setupTasks(
+    project: Project,
+    variant: BaseVariant
+  ) {
+    val copyObfuscationMappingFileTaskProvider = project.tasks.register(
+        "leakCanaryCopyObfuscationMappingFor${variant.name.capitalize()}",
+        CopyObfuscationMappingFileTask::class.java
+    ) {
+      it.mappingFile = getMappingFile(project, variant)
+      it.mergeAssetsDirectory = variant.mergeAssetsProvider.get().outputDir.get().asFile
+
+      val mappingGeneratingTaskProvider =
+        findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
+        ) ?: findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
+        ) ?: throwMissingMinifiedVariantException(project)
+
+      it.dependsOn(mappingGeneratingTaskProvider)
+    }
+
+    getPackageTaskProvider(variant).configure {
+      it.dependsOn(copyObfuscationMappingFileTaskProvider)
+    }
+  }
+
+  private fun getMappingFile(
+    project: Project,
+    variant: BaseVariant
+  ): File {
+    return variant.mappingFile ?: throwMissingMappingFileForVariantException(project, variant.name)
+  }
+
+  private fun findTaskProviderOrNull(
+    project: Project,
+    taskName: String
+  ): TaskProvider<Task>? {
+    return try {
+      project.tasks.named(taskName)
+    } catch (proguardTaskNotFoundException: UnknownTaskException) {
+      null
+    }
+  }
+
+  private fun getPackageTaskProvider(variant: BaseVariant): TaskProvider<out DefaultTask> {
+    return when (variant) {
+      is LibraryVariant -> variant.packageLibraryProvider
+      is ApplicationVariant -> variant.packageApplicationProvider
+      else -> throwNoAndroidPluginException()
+    }
+  }
+
+  private fun throwNoAndroidPluginException(): Nothing {
+    throw GradleException(
+        "LeakCanary deobfuscation plugin can be used only in Android application or library module."
+    )
+  }
+
+  private fun throwMissingMinifiedVariantException(project: Project): Nothing {
+    project.logger.log(
+        DEBUG,
+        """
+          None of the project's variants seem to have minification enabled. 
+          Please make sure that there is at least 1 minified variant in your project.    
+        """.trimIndent()
+    )
+    throwNoAndroidPluginException()
+  }
+
+  private fun throwMissingMappingFileForVariantException(
+    project: Project,
+    variantName: String
+  ): Nothing {
+    project.logger.log(
+        DEBUG,
+        """
+          Missing obfuscation mapping file for $variantName variant.
+          Make sure you used correct variant name in Leak Canary plugin configuration block.
+        """.trimIndent()
+    )
+    throwNoAndroidPluginException()
```

Comment:
I am wondering if the 2 errors are really distinct: as soon as a variant as isMinified = true, it should have a variant mapping file no ?

Reply:
Also, in the code, and maybe for the user, `throwNoAndroidPluginException` seems quite confusing, we should throw a different Gradle error.